### PR TITLE
Decisionlog droprule

### PR DESCRIPTION
**Pull request by mariusblarsen**
We picked up the work at #5053.

This will be used to drop events before they are logged, specified by a drop decision rule.

Fixes: #3945

### CODE REVIEW

**Comment by srenatus** [comment]
Would you mind checking that rebase? It looks like a few unrelated commits have snuck in.

**Comment by srenatus** [comment]
Also thanks for picking this up! 👏

**Comment by johanneslarsson** [comment]
> Would you mind checking that rebase? It looks like a few unrelated commits have snuck in.

We will fix it on Monday

**Comment by mariusblarsen** [comment]
> Would you mind checking that rebase? It looks like a few unrelated commits have snuck in.

The rebase should be fixed now @srenatus, thanks

File: docs/content/management-decision-logs.md [hunk 1]
```diff
@@ -253,6 +253,36 @@ to track **remove** vs **upsert** mask operations.
...
+
+This rule will drop all requests to the _allow_ rule in the _kafka_ package, that returned _true_:
+```ruby
+package system.log
+
+drop {
+  input.path = "kafka/allow"
+  input.result
+}
+```
```
Comment (srenatus) [inline docs/content/management-decision-logs.md:268]: For consistent formatting on the website, let's use another language annotation: 

```
```live:drop_rule_example:module:read_only
package system.log

drop {
  input.path == "kafka/allow"
  input.result
}
```

☝️ Explicit unification (`=`) tends to confuse people, so let's use `==` where possible

File: docs/content/management-decision-logs.md [hunk 2]
```diff
@@ -253,6 +253,36 @@ to track **remove** vs **upsert** mask operations.
...
+
+Log only requests for _delete_ and _alter_ operations (Kafka with opa-kafka-authorizer):
+
+```ruby
```
Comment (srenatus) [inline docs/content/management-decision-logs.md:272]:
```suggestion
```live:drop_rule_example_2:module:read_only
```

File: docs/content/management-decision-logs.md [hunk 3]
```diff
@@ -253,6 +253,36 @@ to track **remove** vs **upsert** mask operations.
...
+package system.log
+
+drop {
+  input.path = "kafka/allow"
+  not input.input.action.operation = ["DELETE", "ALTER"][_]
```
Comment (srenatus) [inline docs/content/management-decision-logs.md:277]: [nit] Let's import `future.keywords.in` and do
```suggestion
  not input.input.action.operation in {"DELETE", "ALTER"}
```

File: plugins/logs/plugin.go [hunk 4]
```diff
@@ -18,6 +18,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/open-policy-agent/opa/topdown/cache"
```
Comment (srenatus) [inline plugins/logs/plugin.go:21]: [nit] Let's put it into the bottom lump of imports

File: plugins/logs/plugin_test.go [hunk 5]
```diff
@@ -1781,6 +1781,99 @@ func TestPluginMasking(t *testing.T) {
...
+			manager, err := plugins.New(
+				nil,
+				"test",
+				store,
+				plugins.EnablePrintStatements(true),
+				plugins.PrintHook(appendingPrintHook{printed: &output}),
+			)
+			if err != nil {
+				t.Fatal(err)
+			} else if err := manager.Start(ctx); err != nil {
```
Comment (srenatus) [inline plugins/logs/plugin_test.go:1847]: [nit] Previous branch stops execution, so we don't need the `else` here
```suggestion
			}
			if err := manager.Start(ctx); err != nil {
```

File: plugins/logs/plugin_test.go [hunk 6]
```diff
@@ -1781,6 +1781,99 @@ func TestPluginMasking(t *testing.T) {
 	}
 }
 
+func TestPluginDrop(t *testing.T) {
+	// Test cases
+	tests := []struct {
+		note       string
+		rawPolicy  []byte
+		errManager error
```
Comment (srenatus) [inline plugins/logs/plugin_test.go:1789]: This appears to be unused, let's drop it, and it's checking code below.

**Review by srenatus — COMMENTED** [review]
This looks pretty good -- thanks again for picking it up. There're a bunch of nitpicks, please bear with me 😄

File: docs/content/management-decision-logs.md (hunk 1 above)
Comment (johanneslarsson) [reply docs/content/management-decision-logs.md:268]: Amen!

File: docs/content/configuration.md [hunk 7]
```diff
@@ -830,7 +830,8 @@ included in the actual bundle gzipped tarball.
 | `decision_logs.reporting.min_delay_seconds` | `int64` | No (default: `300`) | Minimum amount of time to wait between uploads. |
 | `decision_logs.reporting.max_delay_seconds` | `int64` | No (default: `600`) | Maximum amount of time to wait between uploads. |
 | `decision_logs.reporting.trigger` | `string` | No (default: `periodic`) | Controls how decision logs are reported to the remote server. Allowed values are `periodic` and `manual`. |
-| `decision_logs.mask_decision` | `string` | No (default: `system/log/mask`) | Set path of masking decision. |
+| `decision_logs.mask_decision` | `string` | No (default: `/system/log/mask`) | Set path of masking decision. |
+| `decision_logs.drop_decision` | `string` | No (default: `/system/log/drop`) | Set path of drop decision. If this rule evaluates to `true`, the decision will not be logged. |
```
Comment (johanneslarsson) [inline docs/content/configuration.md:834]: Do we need the second sentence, isn't it obvious? "If this rule evaluates to `true`, the decision will not be logged."

File: docs/content/management-decision-logs.md [hunk 8]
```diff
@@ -253,6 +253,36 @@ to track **remove** vs **upsert** mask operations.
 }
 ```
 
+### Dropping decisions
```
Comment (johanneslarsson) [inline docs/content/management-decision-logs.md:256]: Lets stick to case standard used in the docs and perhaps rephrase to "Drop Decision Logs"

**Comment by mariusblarsen** [comment]
Thanks for the instructive feedback, the changes are commited now

**Review by srenatus — APPROVED** [review]
Thanks a lot!

@philipaconrad @anderseknert WDYT?

File: docs/content/management-decision-logs.md [hunk 9]
```diff
@@ -253,6 +253,38 @@ to track **remove** vs **upsert** mask operations.
...
+
+Drop rules filters all decisions, which evaluate to `true`, before logging them.
+
+This rule will drop all requests to the _allow_ rule in the _kafka_ package, that returned _true_:
+```live:drop_rule_example/kafka_allow_rule:module:read_only
+package system.log
+
+drop {
+  input.path == "kafka/allow"
+  input.result
```
Comment (anderseknert) [inline docs/content/management-decision-logs.md:266]: I'd prefer an explicit `input.result == true` here.

**Review by anderseknert — APPROVED** [review]
Nice! Two minor remarks, but LGTM 👍

File: docs/content/management-decision-logs.md (hunk 9 above)
Comment (srenatus) [reply docs/content/management-decision-logs.md:266]: Is that the intention, though? I've read that as "is defined"

Comment (srenatus) [reply docs/content/management-decision-logs.md:266]: My bad. You're right, let's update this.

File: plugins/logs/plugin.go [hunk 10]
```diff
@@ -584,6 +600,16 @@ func (p *Plugin) Log(ctx context.Context, decision *server.Info) error {
 		inputAST:       decision.InputAST,
 	}
 
+	drop, err := p.dropEvent(ctx, decision.Txn, &event)
+	if err != nil {
+		p.logger.Error("Log drop decision failed: %v.", err)
+		return nil
+	}
+
+	if drop {
```
[the remark this answers is not in the document]
Comment (mariusblarsen) [reply plugins/logs/plugin.go:609]: Nice addition, added the debug log here as it seems like the most appropriate place:)

**Review by anderseknert — APPROVED** [review]
LGTM

**Review by srenatus — APPROVED** [review]

**Comment by johanneslarsson** [comment]
@mariusblarsen I think we are missing a commit squash, before this can be merged?

**Comment by johanneslarsson** [comment]
@anderseknert fixed it 👍

**Pull request merged**